### ISSUE TITLE
Get host key without proxying connection

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#3530]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/3542
- Created at 2017-10-09 17:04:32 by [pvolpe](https://pagure.io/user/pvolpe)
- Closed at 2018-07-30 22:06:14 as Fixed
- Assigned to [fidencio](https://pagure.io/user/fidencio)
---

sssd provides sss_ssh_knownhostsproxy which is used as a proxy command for ssh. When invoked it updates /var/lib/sss/pubconf/known_hosts then establishes a connection to the target machine. I would like to get the known key for the target machine and exit without with proxy functionality.  What do you think about adding a flag that either outputs the known key or updates  /var/lib/sss/pubconf/known_hosts and then exits.

### Comments

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2017-10-09 17:13:46](https://pagure.io/SSSD/sssd/issue/3542#471498)**

I think it would be handy also for testing, because without a change like this one we don't have a good way of testing https://github.com/SSSD/sssd/pull/237 for example

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2017-10-09 17:20:18](https://pagure.io/SSSD/sssd/issue/3542#471504)**

> I think it would be handy also for testing, because without a change like this one we don't have a good way of testing https://github.com/SSSD/sssd/pull/237 for example

+1

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2017-10-31 22:32:49](https://pagure.io/SSSD/sssd/issue/3542#476236)**

**Metadata Update from @jhrozek**:
- Issue priority set to: minor
- Issue set to the milestone: SSSD 1.16.1

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2017-12-18 17:24:19](https://pagure.io/SSSD/sssd/issue/3542#485146)**

**Metadata Update from @jhrozek**:
- Issue tagged with: postpone-to-1-16-2

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-01-08 17:28:40](https://pagure.io/SSSD/sssd/issue/3542#487181)**

**Metadata Update from @jhrozek**:
- Issue set to the milestone: SSSD 1.16.2 (was: SSSD 1.16.1)

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-01-08 17:40:15](https://pagure.io/SSSD/sssd/issue/3542#487204)**

**Metadata Update from @jhrozek**:
- Issue **un**tagged with: postpone-to-1-16-2

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-03-20 13:41:17](https://pagure.io/SSSD/sssd/issue/3542#500534)**

**Metadata Update from @jhrozek**:
- Issue priority set to: critical (was: minor)
- Issue set to the milestone: SSSD 2.0 (was: SSSD 1.16.2)

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-04-09 08:14:25](https://pagure.io/SSSD/sssd/issue/3542#504334)**

Does this need more design/discussion or is it "just" a matter of doing it now? From the Cockpit side we can work with either proposal (although I find it a bit more elegant to write to stdout instead of always touching the file system). Thanks!

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-04-09 09:49:23](https://pagure.io/SSSD/sssd/issue/3542#504342)**

I think this just needs someone to hack on the code now.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-06-21 15:33:13](https://pagure.io/SSSD/sssd/issue/3542#517917)**

**Metadata Update from @fidencio**:
- Issue assigned to fidencio

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-07-13 18:14:43](https://pagure.io/SSSD/sssd/issue/3542#521459)**

PR: https://github.com/SSSD/sssd/pull/618

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-07-13 18:14:44](https://pagure.io/SSSD/sssd/issue/3542#521460)**

**Metadata Update from @fidencio**:
- Issue tagged with: PR

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-07-30 22:05:57](https://pagure.io/SSSD/sssd/issue/3542#524238)**

* master:
36f2fe8f6306df3b5495f34110280d0d6133d7b0                                                                                                                                                                                                    
b1141e4141213a43d38d1d22a501a589a7a94956
49bb452049e352655af3bcb354e58e7ee9646274

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-07-30 22:06:15](https://pagure.io/SSSD/sssd/issue/3542#524239)**

**Metadata Update from @jhrozek**:
- Issue close_status updated to: Fixed
- Issue set to the milestone: SSSD 1.16.3 (was: SSSD 2.0)
- Issue status updated to: Closed (was: Open)

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-07-30 22:06:53](https://pagure.io/SSSD/sssd/issue/3542#524240)**

@martinpitt please let me/us know if you'd like to try a test package.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-03 09:52:56](https://pagure.io/SSSD/sssd/issue/3542#524880)**

@jhrozek : Thanks! I will carve out some time to work on this in two weeks (after PTO), so if that feature is in RHEL or Fedora by then, that's fine. Otherwise I would indeed appreciate a test package. That interface looks good indeed, we can use that in cockpit-ssh.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-20 14:52:30](https://pagure.io/SSSD/sssd/issue/3542#526982)**

I haven't tested libssh, but as long as it supports the proxy command, the sss_ssh_knownhostsproxy should work as well.  There is nothing inherently openssh specific (I hope..)

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-21 13:05:38](https://pagure.io/SSSD/sssd/issue/3542#527213)**

Seeing libssh source code there is support for the proxy command, so we should be ok. Thank you.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-23 10:08:10](https://pagure.io/SSSD/sssd/issue/3542#527594)**

@jhrozek : I plan to work on this tomorrow, so if you have a test package (for Fedora 28 preferably, but rhel-8 works too), I'd appreciate. Is that already in sssd 2.0 (which went to  [rawhide](https://koji.fedoraproject.org/koji/buildinfo?buildID=1136070) a week ago? (But this can't just be installed on F28, needs a newer glibc).

I looked at this more closely: We have some cases (particularly, direct URL login to a remote machine, like https://localhost:9090/=user@1.2.3.4/system) where cockpit checks if a host is known even before starting the ssh connection. I think (this was all before my time, and Peter left last year) the concern was to avoid tricking someone into connecting to rogue hosts by making them click on URLs with remote cockpit logins.

So under that aspect, I'm not sure if letting ssh check through ProxyCommand works or is too late. From the ssh POV it should be okay, as ssh then *only* talks through the proxy command and does not make a direct connection to the target host, right? So the question is if sss_ssh_knownhostsproxy also does that, i. e. if it checks the central known hosts database (for a matching host name) before initiating the TCP connection to the target host. If that is so, then we could really just use the ProxyCommand functionality. If not, i. e. it first connects to the target ssh, asks for the host key, and uses that to look up in `known_hosts`, then cockpit-ssh has to do the host name matching directly.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-23 10:43:11](https://pagure.io/SSSD/sssd/issue/3542#527605)**

@martinpitt:

Just replying to the first paragraph: the patch is also part of the latest f28 build (https://koji.fedoraproject.org/koji/buildinfo?buildID=1138209).

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-23 16:15:45](https://pagure.io/SSSD/sssd/issue/3542#527655)**

@martinpitt About the mode of operation, if I understand you correctly, it's the former. sss_ssh_authorizedkeys first looks up the host based on its canonicalized hostname. If the host can't be looked up, sss_ssh_authorizedkeys returns an error, otherwise, if the host is found, it would start proxying the data.

But I was also under the impression that you were after printing the host key only?

btw as @fidencio said, you don't have to use the 2.0 release in rawhide, you can use F-28 as long as you use 1.16.3.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-23 16:38:11](https://pagure.io/SSSD/sssd/issue/3542#527658)**

@jrhozek: That mode of operation would be perfect indeed, then we can use that mode in the future. But teaching all that to cockpit-ssh is a much more intrusive change, and at this point might be too intrusive for RHEL downstream, so we'll use the feature to print the host key for the time being.

Indeed I verified that the change is in latest F28. Thanks!

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-08-31 15:56:46](https://pagure.io/SSSD/sssd/issue/3542#529087)**

@jhrozek : Just letting you know that I now have cockpit-ssh working with the`--pubkey` option. Works very nicely, thank you! I still hope that we can make this more elegant at some day, but this will do nicely for the short term.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-09-01 12:05:20](https://pagure.io/SSSD/sssd/issue/3542#529194)**

You're welcome, thank you for testing. If there are any specific issues you have in mind about making this more elegant, please do file a ticket so we track them somewhere.

---

**Comment from [pvolpe](https://pagure.io/user/pvolpe) at [2018-09-02 12:14:10](https://pagure.io/SSSD/sssd/issue/3542#529240)**

Just thinking aloud here.

@jhrozek: So for now you can either check this with (1) `sss_ssh_knownhostsproxy --pubkey $HOSTNAME` or (2) `ssh-keygen -F $HOSTNAME -f /var/lib/sss/pubconf/known_hosts`.  Both (helper program and cached file location) ought to be an internal implementation detail; (1) is not available in older distributions, but at least the helper program is in `$PATH`. (2) does not work right away as you first have to ensure that the cached file actually  exists, and making the file name location a public API doesn't seem desirable either.

But even *having to know* about this mechanism (and re-implementing it in cockpit-ssh) isn't an elegant solution IMHO, as all of that should be an implementation detail. So in that sense, I don't think we should actually add anything else to `sss_ssh_knownhostsproxy`, the `--pubkey` workaround is good enough for now.

So aside from using `ssh` itself, there might be a more elegant mechanism with using libssh's option parsing and re-using that for known hosts checking. I need to check the new 0.8 API, perhaps it helps with that regard.
